Add Provided Type Converter feature

mzgreen opened the pull request:
This change introduces a new `@ProvidedTypeConverter` annotation that gives developers control over type converter instantiation so they can use external dependencies inside type converters. A type converter class annotated with `@ProvidedTypeConverter` won't be instantiated by Room. Room will expect such converters to be passed at runtime using new `builder.addTypeConverter()` method.

Design doc: https://docs.google.com/document/d/1nmXZUsSlH6WWsT3Duxnkd4LcAPZGcyVCwJe5scnh7HU/edit?usp=sharing
Test: Added `ProvidedTypeConverterTest`
Fixes: b/121067210
RelNote: Room can have provided type converter annotation which allow developers to take control over type converters instantiation

Commit:
Add Type Converter Factory feature

yigit commented on room/compiler/src/main/kotlin/androidx/room/DatabaseProcessingStep.kt:
```diff
@@ -67,7 +68,8 @@ class DatabaseProcessingStep : XProcessingStep {
 
         databases?.forEach { db ->
-            DatabaseWriter(db).write(context.processingEnv)
+            DatabaseWriter(db, context.getTypeConverterFactories(db.element)).write(context
```
you probably want DatabaseProcessor to extract the factory information to the Database value object it creates (instead of parsing here).
the validation (mostly) happens on the Processors.

mzgreen replied:
Sure I'll move it there!

yigit commented on room/compiler/src/main/kotlin/androidx/room/processor/Context.kt:
```diff
@@ -37,7 +38,9 @@ class Context private constructor(
-    private val canRewriteQueriesToDropUnusedColumns: Boolean
+    private val canRewriteQueriesToDropUnusedColumns: Boolean,
+    // currently processed database element to type converter factory map
+    private val typeConverterFactoriesMap: MutableMap<XTypeElement, MutableSet<XType>>
```
i'm not sure why we are keeping this in the context but i need to check more in detail (going into a meeting now :) )

yigit replied:
ok i guess this is here because type converters were initially kept in here because they are scoped to whereever the annotation is used.
If possible, it feels like it would be cleaner if the database processor collected each `used` converter from its DAO's and entities then keep that list in the `Database` value object (to verify later in the generated code that all used ones are provided).

mzgreen replied:
So since it's not possible to give a compile time validation for converter factories, I wanted to perform 2 checks at runtime (when `DatabaseBuilder.build()` is called). Those checks are:

1.  User added `@TypeConverterFactory(FooFactory.class)` on a type converter class but forgot to add the instance of `FooFactory` to database builder.

2. User provided an instance of `FooFactory` to database builder but there is no type converter class annotated with `@TypeConverterFactory(FooFactory.class)`.

In order to do that I needed an information about all `@TypeConverterFactory` annotations to be available in `RoomDatabase` file. I figured out that I can use `DatabaseWriter` to put this info there. `Context` already keeps track of type converters but it's a hierarchical data structure and at `DatabaseProcessor` level I have access only to the top level context. It's not possible to extract this information having access only to this context so I added a "flat" data structure that keeps track of all type converter factories.
Not sure what does it mean that a converter is being `used` or not but currently factories are added to a set inside a context. The logic of adding the factory to this set is in `CustomConverterProcessor` file. I'm not sure if this guarantees that a converter is being `used`. It was the only place I've found that allowed me to implement this feature. I'm open to suggestions though :)

danysantiago replied:
I think we want to move the set of factories to the value object we have to represent a processed database: `androidx.room.vo.Database`. I do still think we need the `context` to gather them since they can be scoped to entity fields, daos, even POJOs so its not easy to gather those from the `DatabaseProcessor`. But what we can do is keep it more local by creating the map of factories in `DatabaseProcessor` and clearing it from the context once we are done and have created a `androidx.room.vo.Database`. It also means we can remove the mutable property `databaseElement` that was added. `DaoWritter` would also stay the same, since the info is in the value object.

mzgreen replied:
I implemented @danysantiago suggestions. Let me know if it's fine now :)

yigit reviewed (commented):
partial review, thanks for sending this!

yigit commented on room/compiler/src/main/kotlin/androidx/room/processor/CustomConverterProcessor.kt:
```diff
@@ -77,29 +78,38 @@ class CustomConverterProcessor(val context: Context, val element: XTypeElement)
+                isKotlinObjectDeclaration || allStatic || constructors.isEmpty() ||
+                        constructors.any {
+                            it.parameters.isEmpty()
+                        }, element, TYPE_CONVERTER_MISSING_NOARG_CONSTRUCTOR
```
we probably want to update this message (in processor errors) to refer to the type converter factory if they can provide the instance.

danysantiago commented on room/common/src/main/java/androidx/room/TypeConverterFactory.java:
```diff
@@ -0,0 +1,38 @@
+     * @return The class of a {@code TypeConverterFactory } that should be used to instantiate a
+     * class that contains the converter methods.
+     */
+    Class<?> value();
```
We can make this `Class<? extends TypeConverterFactory>` if we move the interface to the `common`, make it a top level class (not an inner class of RoomDatabase). Meanwhile to avoid conflict with this annotation, lets make this annotation an inner class of `@TypeConverter` and call it `Factory` so that the usage is then:
```
@TypeConverter.Factory(TimeStampConverterFactory.class)
public static class TimeStampConverter {
```

danysantiago commented on room/runtime/src/main/java/androidx/room/RoomDatabase.java:
```diff
@@ -178,6 +190,27 @@ public void init(@NonNull DatabaseConfiguration configuration) {
+                    configuration.typeConverterFactories.entrySet()) {
+                if (!mTypeConverterFactories.containsKey(entry.getKey())) {
+                    throw new IllegalArgumentException("Unexpected factory " + entry.getKey() + ". "
+                            + "Annotate TypeConverter class with @TypeCoverterFactory annotation "
```
typo in `@TypeCoverterFactory`

danysantiago commented on room/runtime/src/main/java/androidx/room/RoomDatabase.java:
```diff
@@ -178,6 +190,27 @@ public void init(@NonNull DatabaseConfiguration configuration) {
+            if (entry.getValue() == null) {
+                throw new IllegalArgumentException("Missing " + entry.getKey() + " factory "
+                        + "instance. Add it using addTypeConverterFactory method or "
+                        + "remove unnecessaty @TypeConverterFactory annotation from "
```
typo in `unnecessaty`

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/writer/DatabaseWriter.kt:
```diff
@@ -59,11 +65,43 @@ class DatabaseWriter(val database: Database) : ClassWriter(database.implTypeName
     }
 
+    private fun createCreateTypeConverterFactoriesMap():
+            MethodSpec {
```
return type can be in same line as method name

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/writer/DatabaseWriter.kt:
```diff
@@ -59,11 +65,43 @@ class DatabaseWriter(val database: Database) : ClassWriter(database.implTypeName
 
+    private fun createCreateTypeConverterFactoriesMap():
+            MethodSpec {
+        val typeConverterFactoryType = ClassName.get(TypeConverterFactory::class.java)
```
Instead of `ClassName.get(TypeConverterFactory::class.java)` use a constant defined here: https://github.com/androidx/androidx/blob/androidx-master-dev/room/compiler/src/main/kotlin/androidx/room/ext/javapoet_ext.kt#L53

danysantiago commented on room/runtime/src/main/java/androidx/room/RoomDatabase.java:
```diff
@@ -1012,6 +1058,21 @@ private static boolean isLowRamDevice(@NonNull ActivityManager activityManager)
+            if(mTypeConverterFactories == null) {
+                mTypeConverterFactories = new HashMap<>();
+            }
+            mTypeConverterFactories.put(factory.getClass().getCanonicalName(), factory);
```
Since we use string as class identifier names we need to add a proguard rule to avoid obfuscation.
The rule will probably look like this:
`-keepclasseswithmembernames class * implements androidx.room.TypeConverterFactory`
to keep the name of classes that implement the factory interface.

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/writer/DatabaseWriter.kt:
```diff
@@ -59,11 +65,43 @@ class DatabaseWriter(val database: Database) : ClassWriter(database.implTypeName
+            addModifiers(PROTECTED)
+            returns(ParameterizedTypeName.get(CommonTypeNames.MAP, CommonTypeNames.STRING,
+                typeConverterFactoryType))
+            val typeConverterFactoriesVar = "_typeConverterFactoriesMap"
```
Use `val typeConverterFactoriesVar = scope.getTmpVar("_typeConverterFactoriesMap")` to create a local scoped variable name without conflicts.

danysantiago commented on room/runtime/src/main/java/androidx/room/DatabaseConfiguration.java:
```diff
@@ -363,6 +367,59 @@ public DatabaseConfiguration(@NonNull Context context, @Nullable String name,
             @Nullable File copyFromFile,
             @Nullable Callable<InputStream> copyFromInputStream,
             @Nullable RoomDatabase.PrepackagedDatabaseCallback prepackagedDatabaseCallback) {
+        this(context, name, sqliteOpenHelperFactory, migrationContainer, callbacks,
```
This constructor must now be deprecated, add `@Deprecated` and in the method doc add a `@deprecated` tag, see examples above. All other constructor bodies must be updated to call the newest constructor and their `@deprecated` tag updated.

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/solver/types/CustomTypeConverterWrapper.kt:
```diff
@@ -45,13 +50,75 @@ class CustomTypeConverterWrapper(val custom: CustomTypeConverter) :
         }
     }
 
+    fun typeConverterFactory(scope: CodeGenScope, enclosingClassFactory: XType): MethodSpec {
```
Can be private

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/solver/types/CustomTypeConverterWrapper.kt:
```diff
@@ -45,13 +50,75 @@ class CustomTypeConverterWrapper(val custom: CustomTypeConverter) :
+        })
+
+        return scope.writer.getOrCreateMethod(object : ClassWriter.SharedMethodSpec(
+            baseName) {
```
nit: `baseName) {` can be placed in the line before

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/solver/types/CustomTypeConverterWrapper.kt:
```diff
@@ -45,13 +50,75 @@ class CustomTypeConverterWrapper(val custom: CustomTypeConverter) :
+            ): CodeBlock {
+                val methodScope = CodeGenScope(writer)
+                methodScope.builder().apply {
+                    beginControlFlow("if($N == null)", converterField)
```
nit: should produce a space between `if` and `(`

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/solver/types/CustomTypeConverterWrapper.kt:
```diff
@@ -45,13 +50,75 @@ class CustomTypeConverterWrapper(val custom: CustomTypeConverter) :
+                builder: MethodSpec.Builder
+            ) {
+                builder.apply {
+                    addModifiers(Modifier.PRIVATE)
```
add a `Modifier.SYNCHRONIZED` since converter instantiation can race between async queries trying to use it.

danysantiago commented on room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java:
```diff
@@ -0,0 +1,151 @@
+        TestDatabaseWithConverter db = null;
+        Throwable throwable = null;
+        try {
+            db = Room.databaseBuilder(context, TestDatabaseWithConverter.class, "test")
```
Use the `inMemoryDatabaseBuilder` and then you can remove the codes regarding deleting the database file and it isolates the test more. Meanwhile, no need to surround `db` initialization with a try-catch, if something goes wrong the test will fail due to an exception opening or closing the DB, same as what you have now, but its just less noise in the test.

danysantiago commented on room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java:
```diff
@@ -0,0 +1,151 @@
+                    .build();
+            Pet pet = TestUtil.createPet(3);
+            pet.setName("pet");
+            db.getPetDao().insertOrReplace(pet);
```
add a `fail("Show have thrown an IllegalStateException")` at the end of the `try` to make sure we expect an exception

danysantiago commented on room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java:
```diff
@@ -0,0 +1,151 @@
+                db.close();
+            }
+        }
+        Assert.assertThat(throwable, instanceOf(IllegalArgumentException.class));
```
This can be cleaned up a bit by not closing the DB (no need to since it should throw) and asserting the exception in the `catch` block

danysantiago commented on room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java:
```diff
@@ -0,0 +1,151 @@
+
+@MediumTest
+@RunWith(AndroidJUnit4.class)
+public class TypeConverterFactoryTest {
```
Can you please add a test for two different converters relying on the same factory?

mzgreen replied:
I modified the existing one to use 2 converters and single factory. Hope it's fine :)

Commit:
Changes after review

danysantiago commented on room/runtime/src/main/java/androidx/room/RoomDatabase.java:
```diff
@@ -1012,6 +1058,21 @@ private static boolean isLowRamDevice(@NonNull ActivityManager activityManager)
+         */
+        @NonNull
+        public Builder<T> addTypeConverterFactory(@NonNull TypeConverterFactory factory) {
+            if(mTypeConverterFactories == null) {
```
nit: missing space between `if` and `(`

mzgreen replied:
Weird, I ran `./gradlew buildOnServer ktlint` before pushing. It doesn't detect this.

danysantiago replied:
@dlam, does `./gradlew ktlint` perform formatting or is it just checking? I'm also not sure how strict is it with unconventional spaces...

mzgreen replied:
I mean I ran `./gradlew buildOnServer ktlint` as suggested here: https://github.com/androidx/androidx/pull/73
and expected it to fail if there is something wrong. I fixed a lot of formatting issues using this approach before and I believe it detected some missing spaces too. Not sure why it doesn't work in this case. I've just ran it again and got no errors.

dlam replied:
ktlint just checks, ktformat applies the fix

dlam replied:
we also normally run like three different kinds of lint as part of bOS, but I feel ktlint shouldve caught this.

danysantiago commented on room/runtime/src/main/java/androidx/room/RoomDatabase.java:
```diff
@@ -178,6 +190,27 @@ public void init(@NonNull DatabaseConfiguration configuration) {
                     configuration.name);
         }
+
+        if(configuration.typeConverterFactories != null) {
```
nit: missing space between `if` and `(`

danysantiago commented on room/runtime/src/main/java/androidx/room/RoomDatabase.java:
```diff
@@ -178,6 +190,27 @@ public void init(@NonNull DatabaseConfiguration configuration) {
+            }
+        }
+
+        for(Map.Entry<String, TypeConverterFactory> entry: getTypeConverterFactories().entrySet()) {
```
nit: missing space between `for` and `(`

danysantiago commented on room/runtime/src/main/java/androidx/room/RoomDatabase.java:
```diff
@@ -1012,6 +1058,21 @@ private static boolean isLowRamDevice(@NonNull ActivityManager activityManager)
+            if(mTypeConverterFactories == null) {
+                mTypeConverterFactories = new HashMap<>();
+            }
+            mTypeConverterFactories.put(factory.getClass().getCanonicalName(), factory);
```
One interesting issue here is that we build the map with the concrete factory class, so if a user provides a subclass of an expected factory Room won't be able to correctly use it / accept it. One solution to this is processing the factory type and making sure its `final`, i.e. making it a compiler error if its not. This is a bit aggressive IMHO, but would avoid any confusion since we can't quite support it.

mzgreen replied:
Good point, I believe Yigit mentioned this as well. I can add a `final` check but I'm trying to understand why can't we support this. Wouldn't it be possible to do something like this:
1. Try to find the converter in the map and use it if it's there
2. Otherwise try to go over supertypes of a given factory and for every one of them check if it's in the map.
Not sure if it's worth it though.

danysantiago replied:
Right, we can't do this without going through superclasses at runtime via reflection. My take here is, it's not initially worth it, I would prefer if users compose their factories (and delegate) not via class hierarchy. We can always add support for this later (and remove the `final` restriction) if we get users with compelling reasons, but if we don't make it strict now and decided not to support this case in the long run we can't later add the restriction since it would cause ABI incompatibilities.

mzgreen replied:
Makes sense, I'll add `final` check

mzgreen replied:
Do you think we should add an info about `final` requirement in the javadoc for `TypeConverterFactory`  interface?

danysantiago replied:
Yes, definitely

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/processor/DatabaseProcessor.kt:
```diff
@@ -107,7 +107,11 @@ class DatabaseProcessor(baseContext: Context, val element: XTypeElement) {
+                enableForeignKeys = hasForeignKeys,
+                typeConverterFactories = HashSet(context.typeConverterFactories))
+
+        context.typeConverterFactories.clear()
```
This mutable map is still a bit scary, what we can do is move the set to a property in the `DatabaseProcessor` and then let the `Context` have a function that `CustomComnverterProcessor` can call into that is provided by `DatabaseProcessor` that adds to the set.

In other words:
```
class Context (
  ...
  val addTypeConverterFactory: (XType) -> Unit
) {
  ...
  fun fork(..., addTypeConverterFactory: (XType) -> Unit = {  }) {
    ...
  }
}
```
then:
```
class DatabaseProcessor(...) {
    val context = baseContext.fork(
        element = element,
        addTypeConverterFactory = { typeConverterFactories.add(it) }
    )
    private val typeConverterFactories = mutableSetOf<XType>()
}
```

mzgreen replied:
Yeah it's much cleaner! Thanks 👍

mzgreen replied:
There is a slight problem though - it doesn't work 😅 
I think the problem is that `fork` calls `CustomConverterProcessor.findConverters(this, element)` and after that it creates a sub context. So `findConverters` is called before `addTypeConverterFactory` is assigned. I may be wrong though.
Trying to make it work now.

mzgreen replied:
Maybe it could be done by passing `addTypeConverterFactory: (XType) -> Unit` as a third parameter to the `findConverters` function. Just like `element` param.

danysantiago replied:
Passing `addTypeConverterFactory` to `CustomConverterProcessor.findConverters` sounds totally reasonable, then we should make the function property `private` in `Context` to avoid misuse.

mzgreen replied:
Yup! 👍

danysantiago reviewed (changes requested):
This is looking great, just a few more comments, thanks!

Commit:
Changes after review and cleanup

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/processor/CustomConverterProcessor.kt:
```diff
@@ -41,14 +42,19 @@ class CustomConverterProcessor(val context: Context, val element: XTypeElement)
+        fun findConverters(
+            context: Context,
+            element: XElement,
+            addTypeConverterFactory: (XType) -> Unit = {}
```
Lets not have a default value here for `addTypeConverterFactory` it makes sense in the `Context` class but not here since we want to make sure someone gets notified that a type converter factory has to be registered.

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/processor/CustomConverterProcessor.kt:
```diff
@@ -70,7 +76,7 @@ class CustomConverterProcessor(val context: Context, val element: XTypeElement)
     }
 
-    fun process(): List<CustomTypeConverter> {
+    fun process(addTypeConverterFactory: (XType) -> Unit = {}): List<CustomTypeConverter> {
```
If we add a `private val` to `CustomConverterProcessor` in its constructor then we don't need this param since in `findConverters` can build a `CustomConverterProcessor` with the add lambda. This keeps `process()` clean as the processor has been built with everything it needs. It should also not have a default value.

Commit:
Converter processor cleanup

yigit commented on room/runtime/src/main/java/androidx/room/DatabaseConfiguration.java:
```diff
@@ -316,7 +320,64 @@ public DatabaseConfiguration(@NonNull Context context, @Nullable String name,
+                null, null);
+    }
+
+     /**
```
unrealted to this CL but i'm not sure why we are still keeping older configurations.
DatabaseConfiguration is not created by generated code and even though constructor is public, it is `@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP_PREFIX)` . is this because we are expected to keep api compatibilty even for restricted? 
cc: @danysantiago

danysantiago replied:
Yes - It possible the API tools forced me to start doing this some time ago and know they are able to track restricted APIs and more importantly allow us to be incompatible knowing they are restricted but I have not tried. =/

yigit commented on room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java:
```diff
@@ -0,0 +1,162 @@
+            Pet pet = TestUtil.createPet(3);
+            pet.setName("pet");
+            db.getPetDao().insertOrReplace(pet);
+            fail("Show have thrown an IllegalArgumentException");
```
typo: "Should've"

mzgreen replied:
Oops copy paste mistake, thanks 😅

yigit commented on room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java:
```diff
@@ -0,0 +1,162 @@
+            Pet pet = TestUtil.createPet(3);
+            pet.setName("pet");
+            db.getPetDao().insertOrReplace(pet);
+            fail("Show have thrown an IllegalArgumentException");
```
ditto

yigit commented on room/compiler/src/main/kotlin/androidx/room/processor/Context.kt:
```diff
@@ -37,7 +37,8 @@ class Context private constructor(
     val cache: Cache,
-    private val canRewriteQueriesToDropUnusedColumns: Boolean
+    private val canRewriteQueriesToDropUnusedColumns: Boolean,
+    private val addTypeConverterFactory: (XType) -> Unit
```
github doesn't let me comment on the previous discussion so i'll create a new one:
https://github.com/androidx/androidx/pull/75#discussion_r482967139

I'm still worried about this.
We keep type converters in Context because they are scoped to where they are declared and they are stateless.
The current implementation relies on a side effect which is a bit scary especially with caching where we try to avoid re-processing same classes (e.g. https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:room/compiler/src/main/kotlin/androidx/room/processor/TableEntityProcessor.kt;l=51)

I tried to create a repro for that but it didn't fail (not sure why, didn't try harder) but i'm afraid it might fail in similar cases.

In other words, there might be a failure for the following case:
There are two database, `Db1`,`Db2` that share same Entity `E1` where `E1` has a field that declares a type converter.

We start processing `Db1`, visit `E1` and add the factory.
We start processing `Db2`, visit `E1` again but it is cached hence return cached Entity information, missing on the factory.

I don't have a solution yet but the original implementation might be better than having a callback.
That is, have a mechanism for a `subContext` to report to its parent a type adapter.

Now, even that is not perfect though w/ the forks where sometimes we fork a context just to collect errors and sometime surpress them (https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:room/compiler/src/main/kotlin/androidx/room/processor/Context.kt;l=117?q=Context.kt)

Also, it would be great not to report errors on unused type converters but that is a good to have, not that important.

mzgreen replied:
I'm not super familiar with the codebase yet, not sure how exactly caching works but I think I know why it doesn't fail. One of the first things that `DatabaseProcessor` does is forking a baseContext. `fork()` method calls `CustomConverterProcessor.findConverters` before creating subcontext and returning it. So it happens before Daos and Entities are processed. That's why it works - Room executes `findConverters` logic without checking the cache.

I decided to report unused converters because I thought it may help developers to understand why the factory doesn't work even though they added it in the builder. But if you think it's better to not report it then I can remove this check.

yigit replied:
so the case above works because available converters is a cache key.

mzgreen replied:
> so the case above works because available converters is a cache key.

I'm trying to understand this, can you elaborate? Does that mean that E1 will be visited again when processing Db2 because converters is a cache key?

mzgreen replied:
One idea for gathering factories from lower levels (daos/entities etc) would be to modify fork method like this:
```
    fun fork(
        element: XElement,
        forceSuppressedWarnings: Set<Warning> = emptySet(),
        addTypeConverterFactory: ((XType) -> Unit)? = null // make it nullable
    ): Context {
        val suppressedWarnings = SuppressWarningProcessor.getSuppressedWarnings(element)
        val processConvertersResult =
            CustomConverterProcessor.findConverters(this, element, addTypeConverterFactory ?: this.addTypeConverterFactory) // pass argument if not null, otherwise use class property
        val canReUseAdapterStore = processConvertersResult.classes.isEmpty()
        // order here is important since the sub context should give priority to new converters.
        val subTypeConverters = if (canReUseAdapterStore) {
            this.typeConverters
        } else {
            processConvertersResult + this.typeConverters
        }
        val subSuppressedWarnings =
            forceSuppressedWarnings + suppressedWarnings + logger.suppressedWarnings
        val subCache = Cache(cache, subTypeConverters.classes, subSuppressedWarnings)
        val subCanRemoveUnusedColumns = canRewriteQueriesToDropUnusedColumns ||
                element.hasRemoveUnusedColumnsAnnotation()
        val subContext = Context(
                processingEnv = processingEnv,
                logger = RLog(logger.messager, subSuppressedWarnings, element),
                typeConverters = subTypeConverters,
                inheritedAdapterStore = if (canReUseAdapterStore) typeAdapterStore else null,
                cache = subCache,
                canRewriteQueriesToDropUnusedColumns = subCanRemoveUnusedColumns,
                addTypeConverterFactory = addTypeConverterFactory ?: this.addTypeConverterFactory) // same here, use arg if not null, otherwise pass class property
        subContext.databaseVerifier = databaseVerifier
        return subContext
    }
```
That way once a not null callback is passed to fork, it'll be used starting from that level and passed to lower levels. We're creating a proper callbacks that collects factories into a map inside `DatabaseProcessor` so all subcontexts would get this callback.


Still not sure if this covers the caching case with 2 DB sharing same entity.

yigit replied:
OK here is a proposal.
tl;dr; Delegate verification to static methods in generated DAOs 🙈 

This provides the best behavior (we won't validate unless it is used) and does not require any change in the `Context`.

Right now, `CustomTypeConverterWrapper.typeConverterFactory` gets a field from the scope.
We can add a scope function specific to factories which can track all factories that are used.

`DaoWriter`, at the end, can generate a `static` method that returns the list of those factories.

`DatabaseWriter` can just query these static methods to build the list and verify against it. That even allows us to print a better runtime exception where we can tell the user "XXDAo needs this type converter factory".

wdyt?

yigit replied:
> > so the case above works because available converters is a cache key.
> 
> I'm trying to understand this, can you elaborate? Does that mean that E1 will be visited again when processing Db2 because converters is a cache key?

So E1 will be visited again for DB2 and then skipped because we've visited E1.
Right now it doesn't because our caching is not clever enough to understand that we've visited E1 before with same type converters hence it works fine.

yigit commented on room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java:
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2018 The Android Open Source Project
```
can we expand this test a little bit more to have cases where a type converter is declared in a lower layer, like a dao and entity.

for instance, while trying to understand why we need the add factory callback in Context, i removed it and all of these tests still pass since it only tests for factories declared in the database.

yigit replied:
repro for that case:
https://android-review.googlesource.com/c/platform/frameworks/support/+/1419740

mzgreen replied:
I'll look into that, thanks.

mzgreen replied:
Yeah seems like keeping the add factory callback in Context may not be needed. It's only needed to be passed to `fork` method because this method calls `findConverters` logic.

I'll add more tests. But as I'm looking at this now I think you may be right and the tests you mentioned may fail. I need to think about collecting those factories a bit more.

yigit replied:
i have a prototype for the DAO based suggestion, will upload it soon.

yigit replied:
here is the *dirty* prototype for removing it from context fully.
https://android-review.googlesource.com/c/platform/frameworks/support/+/1420426

Basically what i do is:
* add a function to codegen scope that is special for converter factories (ugly)
* in dao, generate a method that will return all of the factory `Class`es (not strings)
* in db, generate a method that will return a `Map<String, List<Class<? : TypeConverterFactory>>` that is a map of DAO name to the list of factories it requires.
* in database constructor, get that list and verify against provided items.

I also made another change to not use the cannonical names as keys, instead, just use assignability.
I'm not sure why we preferred strings so this change might be wrong but i thought it would be handy for cases where a sub-type of the factory will be provided.

Lmk what you think. This is just a prototype, quite a dirty implementation but passes tests.

yigit replied:
cc @danysantiago

mzgreen replied:
Seems like there is not much left to be done here! 😛 Nice work

yigit replied:
it needs some cleanup but yea i'm not unhappy about it. I'm not sure if it is better though, lets see what @danysantiago thinks and then if we'll go with that, i can send a PR to your fork.

You can probably already apply it to here via:
```
git fetch https://android.googlesource.com/platform/frameworks/support refs/changes/26/1420426/2 && git cherry-pick FETCH_HEAD
```
Not sure if that command will require authentication, if not, you just need to run it in your local checkout in theory.

Though I should first clean it up.

danysantiago replied:
These are valid points, I like some of Yigit's approach, the DAO informing the factory it expects and then the Database gathering those for validation. However, I'm not convinced about using the `ClassWritter` and adding type converter factory specific code. I talked with @yigit offline (as in, online via a video call) and he is going to cleanup the usage of `ClassWitter` a bit more by having a more general solution for metadata, which in this case would be factories used. Ultimately, @mzgreen, you should able to cherry-pick his patch, adjust it if necessary and then get this merged.

yigit replied:
here is that change:
https://android-review.googlesource.com/c/platform/frameworks/support/+/1420426/2..4

lmk what you think and i'll clean it up for @mzgreen to cherry-pick.

yigit replied:
ok that implementation is ready for review as well:
https://android-review.googlesource.com/c/platform/frameworks/support/+/1420426

@danysantiago @mzgreen lmk what you prefer on whether we want to review that separately since this one is already reviewed.

The changes in that CL are:
* use class as key, allow subclasses
* collect factories in DAOs
* remove get converter factory map from RoomDatabase but instead provide a getter. This will allow not exposing the map so that it is easier to change in the future if we need to.

yigit committed:
Verify type converter factories in DAOs

This CL changes how we list factories.
Instead of relying on context which is not very reliable as it may not
be used, instead, this CL relies on generated code for figuring out
which factories are used.

As there is no clear path to collect them, I've added a metadata field
to class writer that allows stashing such information. Then each DAO
generates a method that returns the list of them and generated
RoomDatabase collects the list from each dao. This also allows Room
database to provide a better error on where each type converter
factory is used.

I've also changed not to provide the map as a public API but instead,
provide a getTypeConverterFactory method in the room database so that
in the future we can change how it is kept internally.

Bug:121067210
Relnote: You can now provide a TypeConverterFactory for TypeConverters
for Room to use when constructing them. This is especially useful if
the TypeConverter needs some dependencies (e.g. a serializer).

Change-Id: I78a04da9faf87a8d4d2388c4681f465b326d7635

danysantiago commented:
Looks like we are still missing the changes to `MigrationTestHelper` to use the newest non-deprecated constructor from `Configuration`.

Commit:
Fix using deprecated methods in MigrationTestHelper

Commit:
Formatting fixes

Commit:
Use new ProvidedTypeConverter annotation approach instead of introducing TypeConverterFactories

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/processor/Context.kt:
```diff
@@ -127,9 +127,13 @@ class Context private constructor(
     }
 
-    fun fork(element: XElement, forceSuppressedWarnings: Set<Warning> = emptySet()): Context {
+    fun fork(
```
nit: After all the changes looks like this file basically stayed the same and only have some formatting changes, would be good to fully revert it to how it was originally to avoid unnecessary changes in the PR.

mzgreen replied:
Done

danysantiago commented on room/compiler/src/main/kotlin/androidx/room/processor/DatabaseProcessor.kt:
```diff
@@ -37,7 +37,9 @@ import com.squareup.javapoet.TypeName
 
 class DatabaseProcessor(baseContext: Context, val element: XTypeElement) {
-    val context = baseContext.fork(element)
+    val context = baseContext.fork(
```
nit: With the new latest update this file only has a formatting change, would be good to revert it back.

mzgreen replied:
Done

danysantiago commented on room/compiler/src/test/data/databasewriter/output/ComplexDatabase.java:
```diff
@@ -174,6 +177,14 @@ public void clearAllTables() {
+    protected Map<String, List<Class<?>>> getRequiredTypeConverters() {
+        final HashMap<String, List<Class<?>>> _typeConvertersMap =
+                new HashMap<String, List<Class<?>>>();
+        _typeConvertersMap.put("foo.bar.ComplexDao", ComplexDao_Impl.getRequiredConverters());
```
Instead of using DAO class name in Strings we should probably use the Class itself to respect obfuscation.

mzgreen replied:
Right! Changed :)

danysantiago reviewed (changes requested):
Thanks for the latest update, I think this is way better.

Commit:
Use class instead of a canonical name as a map key to respect obfuscation

yigit commented on room/common/src/main/java/androidx/room/ProvidedTypeConverter.java:
```diff
@@ -16,21 +16,15 @@
 /**
- * Implementations of {@code TypeConverterFactory} interface are responsible to instantiate
- * TypeConverters.
+ * Marks a class as a type converter that will be provided to Room at runtime.
```
```suggestion
 * Marks a class as a type converter that will be provided to Room at runtime.
 * If Room uses the annotated type converter class, it will verify that it is provided in the builder and if not, will throw
 * an exception.
```

yigit replied:
i think this got lost.

mzgreen committed:
Update room/common/src/main/java/androidx/room/ProvidedTypeConverter.java

Co-authored-by: Yigit Boyar <yboyar@gmail.com>

Commit:
Fix failing test

Commit:
Add more docs
